File: frontend/src/queries/knowledge.ts
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import { api } from "@/lib/api";

export interface KnowledgeNode {
  id: string;
  organization_id: string;
  node_type: string;
  entity_id?: string;
  name: string;
  properties: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeEdge {
  id: string;
  organization_id: string;
  source_id: string;
  target_id: string;
  edge_type: string;
  weight: number;
  properties: Record<string, unknown>;
  created_at: string;
}

export interface MemoryEntry {
  id: string;
  organization_id: string;
  agent_type: string;
  content: string;
  summary: string;
  importance: number;
  access_count: number;
  last_accessed_at?: string;
  created_at: string;
  updated_at: string;
}

interface SearchResponse<T> {
  data: T[];
  count: number;
  query: string;
}

interface NeighborsResponse {
  nodes: KnowledgeNode[];
  edges: KnowledgeEdge[];
}

export function useSemanticSearch(query: string, type?: string, limit = 10) {
  const params = new URLSearchParams({ q: query, limit: String(limit) });
  if (type) params.set("type", type);

  return useQuery<SearchResponse<KnowledgeNode>>({
    queryKey: ["knowledge", "search", query, type, limit],
    queryFn: () => api.get(`/knowledge/search?${params}`),
    enabled: query.length >= 2,
    staleTime: 30_000,
  });
}

export function useMemorySearch(query: string, limit = 10) {
  const params = new URLSearchParams({ q: query, limit: String(limit) });

  return useQuery<SearchResponse<MemoryEntry>>({
    queryKey: ["knowledge", "memories", query, limit],
    queryFn: () => api.get(`/knowledge/memories?${params}`),
    enabled: query.length >= 2,
    staleTime: 30_000,
  });
}

export function useNodeNeighbors(nodeId: string, depth = 1) {
  return useQuery<NeighborsResponse>({
    queryKey: ["knowledge", "neighbors", nodeId, depth],
    queryFn: () => api.get(`/knowledge/nodes/${nodeId}/neighbors?depth=${depth}`),
    enabled: !!nodeId,
  });
}

export function useAddNode() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (data: { node_type: string; name: string; entity_id?: string; properties?: Record<string, unknown> }) =>
      api.post<KnowledgeNode>("/knowledge/nodes", data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["knowledge"] });
    },
  });
}

export function useAddEdge() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (data: { source_id: string; target_id: string; edge_type: string; weight?: number; properties?: Record<string, unknown> }) =>
      api.post<KnowledgeEdge>("/knowledge/edges", data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["knowledge"] });
    },
  });
}

export function useStoreMemory() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (data: { agent_type: string; content: string; metadata?: Record<string, unknown> }) =>
      api.post<MemoryEntry>("/knowledge/memories", data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["knowledge", "memories"] });
    },
  });
}

File: backend/internal/handler/knowledge.go
package handler

import (
	"strconv"

	"github.com/gofiber/fiber/v3"
	"github.com/google/uuid"

	"github.com/timeless/backend/internal/ai/memory"
	"github.com/timeless/backend/internal/middleware"
)

type KnowledgeHandler struct {
	store *memory.Store
}

func NewKnowledgeHandler(store *memory.Store) *KnowledgeHandler {
	return &KnowledgeHandler{store: store}
}

func (h *KnowledgeHandler) SemanticSearch(c fiber.Ctx) error {
	orgID := middleware.GetOrgID(c)
	query := c.Query("q")
	if query == "" {
		return fiber.NewError(fiber.StatusBadRequest, "Query parameter 'q' is required")
	}

	limitStr := c.Query("limit", "10")
	limit, _ := strconv.Atoi(limitStr)
	if limit <= 0 || limit > 50 {
		limit = 10
	}

	nodeType := c.Query("type")

	nodes, err := h.store.SearchNodes(c.Context(), orgID, query, memory.NodeType(nodeType), limit)
	if err != nil {
		return fiber.NewError(fiber.StatusInternalServerError, "Search failed: "+err.Error())
	}

	return c.JSON(fiber.Map{
		"data":  nodes,
		"count": len(nodes),
		"query": query,
	})
}

func (h *KnowledgeHandler) SearchMemories(c fiber.Ctx) error {
	orgID := middleware.GetOrgID(c)
	query := c.Query("q")
	if query == "" {
		return fiber.NewError(fiber.StatusBadRequest, "Query parameter 'q' is required")
	}

	limitStr := c.Query("limit", "10")
	limit, _ := strconv.Atoi(limitStr)
	if limit <= 0 || limit > 50 {
		limit = 10
	}

	memories, err := h.store.SearchMemories(c.Context(), orgID, query, limit)
	if err != nil {
		return fiber.NewError(fiber.StatusInternalServerError, "Memory search failed: "+err.Error())
	}

	return c.JSON(fiber.Map{
		"data":  memories,
		"count": len(memories),
		"query": query,
	})
}

type AddNodeRequest struct {
	NodeType   string                 `json:"node_type" validate:"required"`
	Name       string                 `json:"name" validate:"required"`
	EntityID   *string                `json:"entity_id"`
	Properties map[string]interface{} `json:"properties"`
}

func (h *KnowledgeHandler) AddNode(c fiber.Ctx) error {
	orgID := middleware.GetOrgID(c)

	var req AddNodeRequest
	if err := c.Bind().JSON(&req); err != nil {
		return fiber.NewError(fiber.StatusBadRequest, "Invalid request body")
	}

	if req.Name == "" || req.NodeType == "" {
		return fiber.NewError(fiber.StatusBadRequest, "name and node_type are required")
	}

	node := &memory.KnowledgeNode{
		OrganizationID: orgID,
		NodeType:       memory.NodeType(req.NodeType),
		Name:           req.Name,
		Properties:     memory.JSON(req.Properties),
	}

	if req.EntityID != nil {
		id, err := uuid.Parse(*req.EntityID)
		if err == nil {
			node.EntityID = &id
		}
	}

	if err := h.store.AddNode(c.Context(), node); err != nil {
		return fiber.NewError(fiber.StatusInternalServerError, "Failed to add node: "+err.Error())
	}

	return c.Status(fiber.StatusCreated).JSON(node)
}

type AddEdgeRequest struct {
	SourceID   string                 `json:"source_id" validate:"required"`
	TargetID   string                 `json:"target_id" validate:"required"`
	EdgeType   string                 `json:"edge_type" validate:"required"`
	Weight     float64                `json:"weight"`
	Properties map[string]interface{} `json:"properties"`
}

func (h *KnowledgeHandler) AddEdge(c fiber.Ctx) error {
	orgID := middleware.GetOrgID(c)

	var req AddEdgeRequest
	if err := c.Bind().JSON(&req); err != nil {
		return fiber.NewError(fiber.StatusBadRequest, "Invalid request body")
	}

	sourceID, err := uuid.Parse(req.SourceID)
	if err != nil {
		return fiber.NewError(fiber.StatusBadRequest, "Invalid source_id")
	}

	targetID, err := uuid.Parse(req.TargetID)
	if err != nil {
		return fiber.NewError(fiber.StatusBadRequest, "Invalid target_id")
	}

	weight := req.Weight
	if weight == 0 {
		weight = 1.0
	}

	edge := &memory.KnowledgeEdge{
		OrganizationID: orgID,
		SourceID:       sourceID,
		TargetID:       targetID,
		EdgeType:       memory.EdgeType(req.EdgeType),
		Weight:         weight,
		Properties:     memory.JSON(req.Properties),
	}

	if err := h.store.AddEdge(c.Context(), edge); err != nil {
		return fiber.NewError(fiber.StatusInternalServerError, "Failed to add edge: "+err.Error())
	}

	return c.Status(fiber.StatusCreated).JSON(edge)
}

func (h *KnowledgeHandler) GetNeighbors(c fiber.Ctx) error {
	nodeID, err := uuid.Parse(c.Params("id"))
	if err != nil {
		return fiber.NewError(fiber.StatusBadRequest, "Invalid node ID")
	}

	depthStr := c.Query("depth", "1")
	depth, _ := strconv.Atoi(depthStr)
	if depth <= 0 || depth > 3 {
		depth = 1
	}

	nodes, edges, err := h.store.GetNeighbors(c.Context(), nodeID, depth)
	if err != nil {
		return fiber.NewError(fiber.StatusInternalServerError, "Failed to get neighbors: "+err.Error())
	}

	return c.JSON(fiber.Map{
		"nodes": nodes,
		"edges": edges,
	})
}

type StoreMemoryRequest struct {
	AgentType string                 `json:"agent_type" validate:"required"`
	Content   string                 `json:"content" validate:"required"`
	Metadata  map[string]interface{} `json:"metadata"`
}

func (h *KnowledgeHandler) StoreMemory(c fiber.Ctx) error {
	orgID := middleware.GetOrgID(c)

	var req StoreMemoryRequest
	if err := c.Bind().JSON(&req); err != nil {
		return fiber.NewError(fiber.StatusBadRequest, "Invalid request body")
	}

	if req.Content == "" || req.AgentType == "" {
		return fiber.NewError(fiber.StatusBadRequest, "content and agent_type are required")
	}

	entry, err := h.store.StoreMemory(c.Context(), orgID, req.AgentType, req.Content, req.Metadata)
	if err != nil {
		return fiber.NewError(fiber.StatusInternalServerError, "Failed to store memory: "+err.Error())
	}

	return c.Status(fiber.StatusCreated).JSON(entry)
}
